packages/core-v2-sdk/src/datasources/ContractDataSource.ts

```diff
@@ -0,0 +1,37 @@
+import { BaseContract } from "ethers";
+import LRUCache from "lru-cache";
+import { CachedDataSource } from "./CachedDataSource";
+
+type AnyFunction = (...args: any) => any;
+
+export type FunctionKeys<T> = Exclude<
+  {
+    [K in keyof T]: T[K] extends AnyFunction ? K : never;
+  }[keyof T],
+  undefined
+>;
```

Comment:
kind of over my head right now, could you leave a comment explaining what this does?